2015/intro_concurrency.slide

```diff
@@ -0,0 +1,375 @@
+
+Read from a channel
+
+   v := <-c // Goroutine blocks until can read
+
+Unbuffered (like a synchronous send/receive)
+
+  make(chan float64)
+
+Buffered (like an asynchronous send/receive)
+
+  make(chan float64, 10)
+
+* Channel communication
+.play intro_concurrency/chan.go /func main/,$
+
+Notice that the order is the same
+
+* Channels
+Can have multiple producers and consumers
+
+.play intro_concurrency/multichan.go /func main/,$
+
+* Range
+The `range` keyword reads from the channel until it is closed
+
+.play intro_concurrency/chan_range.go /func consumer/,$
+
+* Deadlock Detection (not perfect)
+
+.play intro_concurrency/deadlock.go
+
+Race detector available as well
+  
+  go install -race myprogram.go
+
+* Select
+The `select` keyword is a switch statement between channel operations
+
+    select {
+    case v := <- c1:
+    	// Do something
+    case <- c2:
+        // Do something else
+    default:
+        // No action available, do a default action
+    }
+
+If no `default`, goroutine blocks until one of the operations is available
+
+* Select Example - Add values received on channel
+.play intro_concurrency/select.go /func worker/,$
+
+* Adaptive Monte Carlo
+Problem: Find the expected value of a function using Monte Carlo
+
+Catch: Error within a user-specified specified tolerance, *not* specified number of samples
+
+- Keep generating samples until the tolerance is met
+
+Concurrently, of course!
+
+Basic strategy
+- *Worker* goroutines that evaluate function and send results on a channel
+- Collect the results until the tolerance is met
+- Close the channel to stop the workers
+
+* Adaptive Monte Carlo -- Generator
+To make our lives easier, let's define some types
+
+.code intro_concurrency/adaptmc.go /START GENERATOR/,/END GENERATOR/
+
+* Adaptive Monte Carlo -- Worker
+
+.code intro_concurrency/adaptmc.go /START_WORKER/,/END_WORKER/
+
+Each worker generates `sz` samples at a time.
+
+Sends data until receives a signal from `quit`.
+
+* Adaptive Monte Carlo -- Function 
+.code intro_concurrency/adaptmc.go /START EXPECTED/,/END EXPECTED/
+
+* Adaptive Monte Carlo -- Main
+.play intro_concurrency/adaptmc.go /func main/,$
+
+* Distributed memory computation
+No official HPC solution
+
+MPI is the wrong paradigm in many cases (global optimization)
+
+Write a server instead
+
+- Go's originally designed for Google's servers
+
+DEMO (server.go, server_writer.go)
+
+* Distributed memory computation
+But MPI is nice too
+
+MPI-like implementation in Go ([[github.com/btracey/mpi]])
+
+- Experimental
+- Not yet as fast or robust as traditional MPI libraries (uses TCP for communication, some bugs)
+- No special compiler, very simple API 
+    mpi.Send(data interface{}, rank, tag int)
+    mpi.Receive(data interface{}, rank, tag int)
+- Trivial to incorporate shared memory (one process per node instead of per CPU)
+- Implementation about 600 lines of code
+
+DEMO (github.com/btracey/mpi/examples/helloworld)
```

Comment:
I'm happy to remove this slide if you don't think it really fits in. It's there to highlight what's possible, but I don't expect my implementation to become a standard or anything.